scripts/convert_data_format.py

```diff
@@ -0,0 +1,100 @@
+#!/usr/bin/env python3
+"""Convert a recorded pickle file to the new format."""
+import argparse
+import json
+import logging
+import pathlib
+import pickle
+import sys
+
+
+def convert_record_1to2(record):
+
+    del record["num_subjects"]
+    del record["subjectNames"]
+
+    subjects = {}
+    for key in list(record.keys()):
+        if key.startswith("subject_"):
+            subject = record[key]
+            name = subject["name"]
+
+            # delete obsolete stuff
+            del subject["name"]
+            del subject["global_rotation"]["eulerxyz"]
+            del subject["global_rotation"]["helical"]
+            del subject["global_rotation"]["matrix"]
+
+            subjects[name] = subject
+
+            # remove the subject_# entry
+            del record[key]
+
+    record["subjects"] = subjects
+    record["format_version"] = 2
+
+    return record
+
+
+def main():
+    logging.basicConfig(level=logging.INFO)
+
+    parser = argparse.ArgumentParser(description=__doc__)
+    parser.add_argument(
+        "input_file", metavar="FILE", type=pathlib.Path, help="Input file"
+    )
+    parser.add_argument(
+        "output_file", metavar="FILE", type=pathlib.Path, help="Output file."
+    )
+    args = parser.parse_args()
+
+    if args.input_file.suffix == ".json":
+        single_record = True
+        loader = json
+    else:
+        single_record = False
+        loader = pickle
+
+    try:
+        with open(args.input_file, "rb") as f:
+            data = loader.load(f)
+    except FileNotFoundError:
+        logging.fatal("File %s does not exists.", args.input_file)
+        return 1
+    except Exception as e:
+        logging.fatal("Failed to read file %s.  Error: %s", args.input_file, e)
+        return 1
+
+    # if only a single record is loaded, simply wrap it in a list for processing
+    if single_record:
+        data = [data]
+
+    for record in data:
+        if "format_version" not in record:
+            record = convert_record_1to2(record)
+        elif record["format_version"] == 2:
+            logging.info("File is already in the latest format.")
+        else:
+            logging.fatal("Unexpected format version %s", record["format_version"])
+            return 1
+
+    # unwrap
+    if single_record:
+        data = data[0]
```

Comment:
why not keep data as a list of length 1 ? if all files contain a list, we get read of the "single record" special case

Reply:
There are two different file types which are handled by this script.  Pickle files containing a sequence of records (generated by recorder.py) and JSON files which only contain a single record (used for the unit tests, see `tests/data`).